### PR TITLE
Pub smallest update

### PR DESCRIPTION
**Pull request by sigurdm**
(two first commits are a42800e5a2f539dd5d86fdc3a6f3beefc971c753)

This fixes: https://github.com/dependabot/dependabot-core/issues/5391

### CODE REVIEW

**Comment by sigurdm** [comment]
cc @jonasfj

File: pub/lib/dependabot/pub/update_checker.rb [hunk 1]
```diff
@@ -42,26 +42,32 @@ def lowest_resolvable_security_fix_version
...
+        # If the current version is not vulnerable, we stay on it.
+        return version_unless_ignored dependency.version unless vulnerable?
+
+        e = dependency_services_smallest_update
+        return nil if e.nil?
+
+        upgrade = e.find { |u| u["name"] == dependency.name }
+
+        version = upgrade["version"]
+        # TODO: Should we do this upgrade even if the found version is ignored?
```
Comment (jurre) [inline pub/lib/dependabot/pub/update_checker.rb:56]: Ignore conditions also apply to security updates, as per the [docs here](https://docs.github.com/en/code-security/dependabot/dependabot-version-updates/configuration-options-for-the-dependabot.yml-file#ignore):

> Setting this option will also affect pull requests for security updates to the manifest files of this package manager, unless you use `target-branch` to check for version updates on a non-default branch.

(The `target-branch` part is handled outside of core, we only pass in ignore conditions relevant to the branch that's getting updated)

For reference you can also see that we do this for npm [here](https://github.com/dependabot/dependabot-core/blob/main/npm_and_yarn/lib/dependabot/npm_and_yarn/update_checker/latest_version_finder.rb#L71C1-L72)

Comment (sigurdm) [reply pub/lib/dependabot/pub/update_checker.rb]: Thanks for the info. Will remove the TODO :)

**Review by jonasfj — COMMENTED** [review]
LGTM as far as I can see.

**Comment by jurre** [comment]
Thanks! With this change, does [this](https://docs.github.com/en/code-security/dependabot/dependabot-version-updates/configuration-options-for-the-dependabot.yml-file#pub) exception for pub still apply? 🤔

**Comment by sigurdm** [comment]
> Thanks! With this change, does [this](https://docs.github.com/en/code-security/dependabot/dependabot-version-updates/configuration-options-for-the-dependabot.yml-file#pub) exception for pub still apply? 🤔

I believe it still applies. This PR only touches security updates, not handling of ignored versions.

**Review by jurre — APPROVED** [review]

File: pub/lib/dependabot/pub/update_checker.rb [hunk 2]
```diff
@@ -42,26 +42,31 @@ def lowest_resolvable_security_fix_version
...
       end
 
       def updated_requirements
         # Requirements that need to be changed, if obtain:
-        # latest_resolvable_version
-        entry = current_report["singleBreaking"].find { |d| d["name"] == dependency.name }
+        # latest_resolvable_version or lowest_security_fix_version
+        entry = if vulnerable?
+                  updates = dependency_services_smallest_update
+                  raise "Cannot upgrade from vulnerability without unlock." if update.size > 1
```
Comment (deivid-rodriguez) [inline pub/lib/dependabot/pub/update_checker.rb:64]: Tried this against a repo and run into an error due to a typo here:

```suggestion
                  raise "Cannot upgrade from vulnerability without unlock." if updates.size > 1
```

I guess we should get this covered, maybe through a smoke test at https://github.com/dependabot/smoke-tests.

Comment (sigurdm) [reply pub/lib/dependabot/pub/update_checker.rb]: Thanks for catching this!

Will you be able to set up a smoke test?

Comment (deivid-rodriguez) [reply pub/lib/dependabot/pub/update_checker.rb]: Not right now, but I'll try set that up!

**Comment by deivid-rodriguez** [comment]
I retried the update after the typo fix and (in retrospective, as expected) I got the error:

> Cannot upgrade from vulnerability without unlock.

This was unexpected to me since the "highest version fix" worked fine. One potential cause is that the update I'm trying is from https://pub.dev/packages/http/versions/0.13.2 to https://pub.dev/packages/http/versions/0.13.3. That actually _removes_ one subdependency, so maybe that's showing as two updates, but it's a case that should be fine?

**Comment by sigurdm** [comment]
> That actually removes one subdependency, so maybe that's showing as two updates, but it's a case that should be fine?

Yeah, that is a good point.

Should we make it that we only complain when the updates contains more than one change to the **manifest** (as opposed to changes in transitive dependencies that affect only the lockfile)? Or should we count the number of non-deleting changes?

I guess the reason we put a limit, is that security updates should be somewhat minimal...

**Comment by deivid-rodriguez** [comment]
We try minimal security updates in terms of version moves, that's why we implement "lowest fixed version" logic, but it's more important that the security update can be fixed, even if it involves unlocking other dependencies. For example, I believe for NPM we have some logic to unlock parent dependencies if that's necessary to get out of a security vulnerability in a transitive dependency.

With that in mind, I'd go with whatever solution allows us to fix more security alerts.

**Comment by sigurdm** [comment]
Ah, now I remember why I inserted the check - the api of `updated_requirements` only allows us to return a single update, as opposed to `updated_dependencies_after_full_unlock` that returns a list of upgraded requirements.

I'll make the check "smarter" to only count breaking changes...

**Comment by deivid-rodriguez** [comment]
By breaking changes, you mean major bumps in transitive dependencies? If that's necessary to fix a vulnerability, I think it's not a problem.

I believe it's always best to show users a path forward (even if the proposed PR needs more work in terms of breaking changes) than just saying "this alert cannot be fixed"?

**Comment by sigurdm** [comment]
> By breaking changes, you mean major bumps in transitive dependencies? If that's necessary to fix a vulnerability, I think it's not a problem.

No, I meant bumps that require an unlock in the manifest - sorry that was unclear.

**Comment by deivid-rodriguez** [comment]
Interesting. So Pub does have the ability to unlock other direct dependencies to get a security upgrade done. We should eventually try to make that work too then, but this seems like a good start! Will give the updated PR a try, thanks for your work!

**Comment by deivid-rodriguez** [comment]
Sorry for dropping the ball (again). I'll try this again and hopefully ship it tomorrow.

**Comment by sigurdm** [comment]
NP - just didn't want this to fall between the cracks :)

**Comment by deivid-rodriguez** [comment]
Hei @sigurdm!

I tried again and I'm still getting the error "Cannot upgrade from vulnerability without unlocking other packages." on an alert where we can now generate a proper PR. I'll have a closer look when I can.

It's just a `pubspec.lock` that has `http` locked to "0.13.2" (vulnerable to https://github.com/advisories/GHSA-4rgh-jx4f-qfcq). Right now we create an update to "0.13.5", the expected outcome of this PR is to generate an update to "0.13.3".

**Comment by sigurdm** [comment]
Hmm, not sure what goes wrong.

I created a pubspec.lock locked to http 0.13.2, and ran:

```
SECURITY_ADVISORIES='[{"dependency-name":"http","affected-versions":["0.13.2"]}]' bin/dry-run.rb pub sigurdm/private_repo2 --security-updates-only
....
 => bump http from 0.13.2 to 0.13.3
....
```

Is there a way I can replicate what you see?

**Comment by deivid-rodriguez** [comment]
I guess the repo I was using is not as simple as I said! Try with https://github.com/dsp-testing/security-updates-for-pub.

**Pull request merged**